pkg/apis/meta/v1/unstructured/converter.go

```diff
@@ -0,0 +1,204 @@
+			data, err := json.Marshal(sv.Interface().(string))
+			if err != nil {
+				return fmt.Errorf("Error encoding string to json")
+			}
+			ret := unmarshal.Call([]reflect.Value{reflect.ValueOf(data)})[0].Interface()
+			if ret != nil {
+				return ret.(error)
+			}
+			return nil
+		}
+	}
+
+	switch dt.Kind() {
+	case reflect.Map:
+		return c.mapFromUnstructured(sv, dv)
+	case reflect.Slice:
+		return c.sliceFromUnstructured(sv, dv)
+	case reflect.Ptr:
+		return c.pointerFromUnstructured(sv, dv)
+	case reflect.Struct:
+		return c.structFromUnstructured(sv, dv)
+	case reflect.Interface:
+		return c.interfaceFromUnstructured(sv, dv)
+	default:
+		return fmt.Errorf("Unrecognized type: %v", dt.Kind())
+	}
+}
+
+func fieldNameFromField(field *reflect.StructField) string {
+	jsonTag := field.Tag.Get("json")
+	if len(jsonTag) == 0 {
+		// FIXME: This should start with small letter.
+		return field.Name
+	}
+	return strings.Split(jsonTag, ",")[0]
+}
+
+func unwrapInterface(v reflect.Value) reflect.Value {
+	for v.Kind() == reflect.Interface {
+		v = v.Elem()
+	}
+	return v
+}
+
+func (c *Converter) mapFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() != reflect.Map {
+		return fmt.Errorf("Cannot restore map from %v", st.Kind())
+	}
+
+	if !st.Key().AssignableTo(dt.Key()) && !st.Key().ConvertibleTo(dt.Key()) {
+		return fmt.Errorf("Cannot copy map with non-assignable keys: %v %v", st.Key(), dt.Key())
+	}
+
+	if sv.IsNil() {
+		dv.Set(reflect.Zero(dt))
+		return nil
+	}
+	dv.Set(reflect.MakeMap(dt))
+	for _, key := range sv.MapKeys() {
+		value := reflect.New(dt.Elem()).Elem()
+		if err := c.fromUnstructured(sv.MapIndex(key), value); err != nil {
+			return err
+		}
+		if st.Key().AssignableTo(dt.Key()) {
+			dv.SetMapIndex(key, value)
+		} else {
+			dv.SetMapIndex(key.Convert(dt.Key()), value)
+		}
+	}
+	return nil
+}
+
+func (c *Converter) sliceFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() != reflect.Slice {
+		return fmt.Errorf("Cannot restore slice from %v", st.Kind())
+	}
+
+	if sv.IsNil() {
+		dv.Set(reflect.Zero(dt))
+		return nil
+	}
+	dv.Set(reflect.MakeSlice(dt, sv.Len(), sv.Cap()))
+	for i := 0; i < sv.Len(); i++ {
+		if err := c.fromUnstructured(sv.Index(i), dv.Index(i)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Converter) pointerFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+
+	if sv.IsNil() {
+		dv.Set(reflect.Zero(dt))
+		return nil
+	}
+	dv.Set(reflect.New(dt.Elem()))
+	switch st.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		return c.fromUnstructured(sv.Elem(), dv.Elem())
+	default:
+		return c.fromUnstructured(sv, dv.Elem())
+	}
+}
+
+func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() != reflect.Map {
+		return fmt.Errorf("Cannot restore struct from: %v", st.Kind())
+	}
+
+	for i := 0; i < dt.NumField(); i++ {
```

Comment:
I'm thinking if necessary later on we can create an intermediate representation of these traversals to make  some of the checks unnecessary.